data/nat/choose.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2018 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+Mostly based on Jeremy Avigad's choose file in lean 2
+-/
+
+
+import data.nat.basic
+
+open nat
+def choose : ℕ → ℕ → ℕ
+| _              0  := 1
+| 0        (succ k) := 0
+| (succ n) (succ k) := choose n k + choose n (succ k)
+
+@[simp]
+lemma choose_zero_right (n : ℕ) : choose n 0 = 1 := by cases n; refl
+
+@[simp]
+lemma choose_zero_succ (k : ℕ) : choose 0 (succ k) = 0 := rfl
+
+lemma choose_succ_succ (n k : ℕ) : choose (succ n) (succ k) = choose n k + choose n (succ k) := rfl
+
+lemma choose_eq_zero_of_lt {n : ℕ} : ∀ {k}, n < k → choose n k = 0 :=
+begin
+  induction n with n hi,
+  { assume k hk, cases k,
+    exact absurd hk dec_trivial,
+    exact choose_zero_succ _ },
+  { assume k hk,
+    cases k with k',
+    { exact absurd hk dec_trivial },
+    { rw [choose_succ_succ, hi (lt_of_succ_lt_succ hk), hi (lt_of_succ_lt hk)] } }
+end
+
+@[simp]
+lemma choose_self (n : ℕ) : choose n n = 1 :=
+by induction n; simp [*, choose, choose_eq_zero_of_lt (lt_succ_self _)]
+
+@[simp]
+lemma choose_succ_self (n : ℕ) : choose n (succ n) = 0 := choose_eq_zero_of_lt (lt_succ_self _)
+
+@[simp]
+lemma choose_one_right (n : ℕ) : choose n 1 = n :=
+by induction n; simp [*, choose]
+
+lemma choose_pos {n : ℕ} : ∀ {k}, k ≤ n → 0 < choose n k :=
+begin
```

Comment:
this may also be a good application for the equation compiler. I guess 3 cases (`0 n`, `(n+1) 0`, `(n+1) (k+1)`) are needed, but they could go through using `simp`.